pkg/fsp/header.go

```diff
@@ -0,0 +1,231 @@
+	FSPSiliconInitEntryOffset uint32
+}
+
+// Summary prints a multi-line summary of the header's content.
+func (ih InfoHeaderRev3) Summary() string {
+	s := fmt.Sprintf("Signature                   : %s\n", ih.Signature)
+	s += fmt.Sprintf("Header Length               : %d\n", ih.HeaderLength)
+	s += fmt.Sprintf("Reserved1                   : %#04x\n", ih.Reserved1)
+	s += fmt.Sprintf("Spec Version                : %s\n", ih.SpecVersion)
+	s += fmt.Sprintf("Header Revision             : %d\n", ih.HeaderRevision)
+	s += fmt.Sprintf("Image Revision              : %s\n", ih.ImageRevision)
+	s += fmt.Sprintf("Image ID                    : %s\n", ih.ImageID)
+	s += fmt.Sprintf("Image Size                  : %#08x %d\n", ih.ImageSize, ih.ImageSize)
+	s += fmt.Sprintf("Image Base                  : %#08x %d\n", ih.ImageBase, ih.ImageBase)
+	s += fmt.Sprintf("Image Attribute             : %s\n", ih.ImageAttribute)
+	s += fmt.Sprintf("Component Attribute         : %s\n", ih.ComponentAttribute)
+	s += fmt.Sprintf("Cfg Region Offset           : %#08x %d\n", ih.CfgRegionOffset, ih.CfgRegionOffset)
+	s += fmt.Sprintf("Cfg Region Size             : %#08x %d\n", ih.CfgRegionSize, ih.CfgRegionSize)
+	s += fmt.Sprintf("Reserved2                   : %#08x\n", ih.Reserved2)
+	s += fmt.Sprintf("TempRAMInit Entry Offset    : %#08x %d\n", ih.TempRAMInitEntryOffset, ih.TempRAMInitEntryOffset)
+	s += fmt.Sprintf("Reserved3                   : %#08x\n", ih.Reserved3)
+	s += fmt.Sprintf("NotifyPhase Entry Offset    : %#08x %d\n", ih.NotifyPhaseEntryOffset, ih.NotifyPhaseEntryOffset)
+	s += fmt.Sprintf("FSPMemoryInit Entry Offset  : %#08x %d\n", ih.FSPMemoryInitEntryOffset, ih.FSPMemoryInitEntryOffset)
+	s += fmt.Sprintf("TempRAMExit Entry Offset    : %#08x %d\n", ih.TempRAMExitEntryOffset, ih.TempRAMExitEntryOffset)
+	s += fmt.Sprintf("FSPSiliconInit Entry Offset : %#08x %d\n", ih.FSPSiliconInitEntryOffset, ih.FSPSiliconInitEntryOffset)
+	return s
+}
+
+// ImageRevision is the image revision fielf of the FSP info header.
+type ImageRevision uint32
+
+func (ir ImageRevision) String() string {
+	return fmt.Sprintf("%d.%d.%d.%d",
+		(ir>>24)&0xff,
+		(ir>>16)&0xff,
+		(ir>>8)&0xff,
+		ir&0xff,
+	)
+}
+
+// SpecVersion represents the spec version as a packed BCD two-digit,
+// dot-separated unsigned integer.
+type SpecVersion uint8
+
+func (sv SpecVersion) String() string {
+	return fmt.Sprintf("%d.%d", (sv>>4)&0x0f, sv&0x0f)
+}
+
+// ImageAttribute represents the image attributes.
+type ImageAttribute uint16
+
+func (ia ImageAttribute) String() string {
+	ret := fmt.Sprintf("%#04x GraphicsDisplaySupported", uint16(ia))
+	if uint16(ia) & ^(uint16(1)) != 0 {
+		ret += " (reserved bits are not zeroed)"
+	}
+	return ret
+}
+
+// ComponentAttribute represents the component attribute.
+type ComponentAttribute uint16
+
+func (ca ComponentAttribute) String() string {
+	var attrs []string
+	if uint16(ca)&0x01 == 1 {
+		attrs = append(attrs, "ReleaseBuild")
```

Comment:
Note that you can split this up into different functions to get the attributes, if you ever want to kinda like how file.go does it. Might be worth putting a TODO here.

Reply:
this is already an attribute-specific method if that's what you mean. Otherwise can you expand on it?

Reply:
I mean it more like how fileAttr is handled in pkg/uefi/file.go. The attributes field is a one byte field, but there are separate functions for IsLarge etc...

Reply:
This would make it easier to unit test specific attributes.

Reply:
oh I see - basically just attaching methods that test the flag bits

Reply:
done